ratelimit_test.go

```diff
@@ -47,95 +79,69 @@ func TestUnlimited(t *testing.T) {
 }
 
 func TestRateLimiter(t *testing.T) {
-	var wg sync.WaitGroup
-	wg.Add(1)
-	defer wg.Wait()
-
-	clock := clock.NewMock()
-	rl := ratelimit.New(100, ratelimit.WithClock(clock), ratelimit.WithoutSlack)
-
-	count := atomic.NewInt32(0)
-
-	// Until we're done...
-	done := make(chan struct{})
-	defer close(done)
-
-	// Create copious counts concurrently.
-	go job(rl, count, done)
-	go job(rl, count, done)
-	go job(rl, count, done)
-	go job(rl, count, done)
-
-	clock.AfterFunc(1*time.Second, func() {
-		assert.InDelta(t, 100, count.Load(), 10, "count within rate limit")
-	})
-
-	clock.AfterFunc(2*time.Second, func() {
-		assert.InDelta(t, 200, count.Load(), 10, "count within rate limit")
-	})
-
-	clock.AfterFunc(3*time.Second, func() {
-		assert.InDelta(t, 300, count.Load(), 10, "count within rate limit")
-		wg.Done()
+	runTest(t, func(r runner) {
+		rl := ratelimit.New(100, ratelimit.WithClock(r.clock), ratelimit.WithoutSlack)
+
+		// Create copious counts concurrently.
+		r.job(rl)
+		r.job(rl)
+		r.job(rl)
+		r.job(rl)
+
+		r.clock.AfterFunc(1*time.Second, func() {
+			assert.InDelta(t, 100, r.count.Load(), 10, "count within rate limit")
+		})
+
+		r.clock.AfterFunc(2*time.Second, func() {
+			assert.InDelta(t, 200, r.count.Load(), 10, "count within rate limit")
+		})
+
+		r.wg.Add(1)
```

Comment:
a test function being able to modify the state of the runner directly feels a little odd. would it be a better idea to pass in a runner interface instead? exposing the waitgroup directly allows for crazy things like calling wg.Wait inside the test function which can cause a deadlock. not that we expect people to do that, but it's better to reduce the API surface to avoid accidental problems

Reply:
Hm.

Agree that having a smaller surface is a good thing, but not sure if this won't over-complicate things here. Like,
- there are literally two tests (one of which tests nothing, I think)
- we're unlikely to have more tests
- if we cause a deadlock than it's on us ;)

Maybe I should have called the `runner` just `args` - would have been closer to my goal of just reducing the noise.

If I pass in the interface I also have to take away access to `r.count`, `r.doneCh` (test below) and I'll need to pass in a function as an argument to `run` (since TestDelayedRateLimiter) is doing some custom work below.

Anyway, can take a stab, why not.

Reply:
Okay, so if we passed in the interface we kinda need more methods:

```
type runner interface {
	// job tries to Take() on passed in limiter in a loop/goroutine.
	job(rls ...ratelimit.Limiter)
	// waitAssert asserts given count matches after some time.
	waitAssert(t time.Duration, count int)
	// getClock return fake clock.
	getClock() ratelimit.Clock
	// add adds time to the fake clock
	add(t time.Duration)
	// afterFunc executes a func after some time.
	afterFunc(t time.Duration, fn func())
}
```

It's a bit verbose ... but I'm starting to like it - the actual test code looks quite clear.

Reply:
To be fair, I don't think we'll need `afterFunc` once we fix TestDelayed ... to make sense.

But for now it should stay here.